programs/psa/psa_hash.c

```diff
@@ -0,0 +1,162 @@
+ *  one-shot API call psa_hash_compute() and the using multi-part
+ *  operation, which requires psa_hash_setup(), psa_hash_update() and
+ *  psa_hash_finish(). The multi-part operation is popular on embedded
+ *  devices where a rolling hash needs to be computed.
+ *
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+
+#include "psa/crypto.h"
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "mbedtls/build_info.h"
+#include "mbedtls/platform.h"
+
+/* The algorithm used by this demo is SHA 256.
+ * Please see include/psa/crypto_values.h to see the other
+ * algorithms that are supported. If you switch to a different
+ * algorithm you will need to update the hash data in the
+ * SAMPLE_HASH_DATA macro below.*/
+
+#define HASH_ALG PSA_ALG_SHA_256
+
+#define SAMPLE_HASH_DATA {                                                      \
+        0x7f, 0x83, 0xb1, 0x65, 0x7f, 0xf1, 0xfc, 0x53, 0xb9, 0x2d, 0xc1, 0x81, \
+        0x48, 0xa1, 0xd6, 0x5d, 0xfc, 0x2d, 0x4b, 0x1f, 0xa3, 0xd6, 0x77, 0x28, \
+        0x4a, 0xdd, 0xd2, 0x00, 0x12, 0x6d, 0x90, 0x69 \
+}
+
+const uint8_t sample_message[] = "Hello World!";
+
+const uint8_t sample_hash[] = SAMPLE_HASH_DATA;
+const size_t sample_hash_len = sizeof(sample_hash);
+
+#if !defined(MBEDTLS_PSA_CRYPTO_C) || !defined(PSA_WANT_ALG_SHA_256)
+int main(void)
+{
+    mbedtls_printf("MBEDTLS_PSA_CRYPTO_C and PSA_WANT_ALG_SHA_256"
+                   "not defined.\r\n");
+    return EXIT_SUCCESS;
+}
+#else
+
+int main(void)
+{
+    psa_status_t status;
+    uint8_t hash[PSA_HASH_LENGTH(HASH_ALG)];
+    size_t hash_length;
+    psa_hash_operation_t hash_operation = PSA_HASH_OPERATION_INIT;
+    psa_hash_operation_t cloned_hash_operation = PSA_HASH_OPERATION_INIT;
+
+    mbedtls_printf("PSA Crypto API: SHA-256 example\n\n");
+
+    status = psa_crypto_init();
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("psa_crypto_init failed\n");
+        return EXIT_FAILURE;
+    }
+
+    /* Compute hash using multi-part operation */
+
+    status = psa_hash_setup(&hash_operation, HASH_ALG);
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("psa_hash_setup failed\n");
+        return EXIT_FAILURE;
+    }
+
+    /* Note: Here we use sizeof(sample_message) - 1 since we don't wish to
+     * include the null byte in the hash computation */
```

Comment:
Any reason why not using `strlen()`? We're clearly assuming that the data is null-terminated...

Reply:
No, none at all, that is an oversight...

Reply:
No, this should remain `sizeof`. We are not assuming that the data is null-terminated, and we should not make such an assumption in sample code. This code works with binary data. You can edit the source file and put a null byte in `sample_message` and the code will still work. The only subtlety is removing the _final_ null byte added by the C language.

Reply:
I actually agree and was going to write something to this effect after looking at it more closely. I have not changed this in the final version, which is now ready for review.

Reply:
I'm sorry, I disagree. Here we have a `- 1` which is important to get the actual expected hash value, which is required because of the way `sample_message[]` is declared and initialised. That is close coupling, at a distance.

Either the message should be `char sample_message[]` and we use `strlen()`, or the message should be properly initialised, without the trailing null terminated.

> You can edit the source file and put a null byte in sample_message and the code will still work

No, it won't. You also have to change this line here, or the hash value computed will be wrong.

> The only subtlety is removing the final null byte added by the C language.

I don't think that is acceptable in a sample program

Reply:
Hashes are very commonly used on binary data, so our sample code should be able to cope with binary data. Using `strlen` is likely to lead to a real bug where a null byte hides data from the hash calculation.

Since we can't change the C language, I think the best we can do is to arrange for the array to have the correct length:

```
/* Note that we hard-code the length of the message in this example, because
 * if we let the compiler calculate it (uint8_t sample_message[] = ...), then
 * sample_message would contain a terminating null byte, which would then be
 * included in the hash calculation. */
const uint8_t sample_message[12] = "Hello World!";
```

And then we can use `sizeof(sample_message)` as the message size.

Alternatively, we could define
```
const uint8_t sample_message[] = "Hello World!";
/* sample_message has a terminating null byte, which is not part of the
 * message, so we make sure to subtract it to get the message length. */
const size_t sample_message_size = sizeof(sample_message) - 1;
```

Reply:
I am happy to continue to use `sizeof()` if we move its use adjacent to the definition of `sample_message`, so I prefer the second suggestion (under "Alternatively, we could define")